Shipping rates UI changes: Make Setup MC support country-level free shipping and minimum order amount

ecgan opened the pull request:
### Changes proposed in this Pull Request:

- Part of a PR refactor from the huge PR https://github.com/woocommerce/google-listings-and-ads/pull/1218.
- Depends on PR:
  - https://github.com/woocommerce/google-listings-and-ads/pull/1266
  - https://github.com/woocommerce/google-listings-and-ads/pull/1262
    - https://github.com/woocommerce/google-listings-and-ads/pull/1202
    - https://github.com/woocommerce/google-listings-and-ads/pull/1249
- Partly addresses the tech debt: https://github.com/woocommerce/google-listings-and-ads/issues/288.

Figma: 2bb94-pb

This PR contains changes to make the Edit Free Listings support country-level free shipping and minimum order amount setup. It uses the `ShippingRateSection` shared component created from #1266.

#### Prior to this PR

In **Setup MC**, the shipping rate code has the following file structure:

![image](https://user-images.githubusercontent.com/417342/154834478-d82878e8-8bb6-48d9-a940-be7a6f15375c.png)

There are a number of "smart components" where the components have the logics the make API calls. For example, see `AddRateModal`: https://github.com/woocommerce/google-listings-and-ads/blob/a6963bc069d58bb0ebf361ae41f9a1633de0b4da/js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/add-rate-button/add-rate-modal/index.js#L45-L55

Upon clicking "Save" button, it will make API call to upsert shipping rate, and update wp.data store, and the UI would update accordingly.

One reason for the above implementation is that we want to have an "auto-save" feature for the Setup MC flow. If users were to proceed through the Setup MC flow and quit halfway, they can come back to it later and continue from where they left off.

#### In this PR

We make use of the `ShippingRateSection` shared component, so all the interactions within ShippingRateSection is the same as the Edit Free Listings flow. Shipping rates data are retrieved and stored in the parent form state (with the form key `shipping_country_rates`), and are propagated up and down to the components via `value` and `onChange` props.

The auto-save feature is now done in `FormContent` using `useAutoSaveShippingRatesEffect`, just like `useAutoSaveSettingsEffect`.

### Screenshots:


Shipping Rates section, with the new "offer free shipping" card and "minimum order" card. This UI and the code is the same for both Setup MC flow and Edit Free Listings flow.

![image](https://user-images.githubusercontent.com/417342/154556792-942a6d6e-9254-4bb6-8c3c-6662a956672a.png)

📹  Demo video with my voice annotation:

https://user-images.githubusercontent.com/417342/154854800-d3d9a1a7-d93e-4ebf-b807-520e4ea5fb11.mov

### Detailed test instructions:

Note: 

1. Open the browser dev tools network panel to observe the API calls.

#### Setup MC

Since the shipping rate section is essentially the same as Edit Free Listings, the steps to test the shipping rate section is the same as above in Edit Free Listings. See PR #1266.

Here, the following test instructions will focus to test the features specific to Setup MC.

1. Make sure that you already have some existing shipping zones settings setup in WooCommerce settings (`/wp-admin/admin.php?page=wc-settings&tab=shipping`). This will be used to test shipping rates suggestions below.
2. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc`. Complete Step 1 account connections.
3. In Step 2, the target audience countries should be auto-populated based on your shipping zones settings. (This is existing feature, out of this PR.)
4. In Step 3, select "flat" shipping rates. Shipping rates should be auto-populated based on shipping rates suggestions.
5. Test the shipping rates section as per above Edit Free Listings instructions.
6. To test the auto-save feature, reload the page while you are in Step 3. The values before and after the reload should be the same.
7. Go back to Step 2, remove a country from the target audience, and go to Step 3. Your shipping rates would have been reset, and then re-populated with shipping rates suggestions.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

ecgan committed:
Move useShippingRatesWithSavedSuggestions files, to be used in SetupFreeListings.

ecgan committed:
Use FlatShippingRatesInputCards in Setup MC, replacing ShippingRateSetup.

ecgan committed:
Merge branch 'tweak/return-rate-after-upsert' into feature/shipping-rates-api-ui-changes

ecgan committed:
Change upsertShippingRates based on new API changes.

New API changes is from https://github.com/woocommerce/google-listings-and-ads/pull/1249. Upsert now returns shipping rates objects with id.

ecgan committed:
Add saveShippingRates and automatically save shipping rates in Setup MC.

ecgan committed:
Use useDebouncedCallbackEffect in useAutoSaveShippingRatesEffect.

ecgan committed:
Use useIsEqualRefValue in useDebouncedCallbackEffect.

This is so that the effect is not triggered when value is still equal.

ecgan committed:
Change first render useRef in useDebouncedCallbackEffect.

This is because if we were to assign value to current, and if value is falsey value, it may not work as expected.

We change ref to boolean for better clarity.

ecgan committed:
Auto save settings when settings are changed, not when shipping rates are changed.

ecgan committed:
Better implementation for saveShippingRates.

The old implementation was deleting all the old shipping rates and upserting new ones, causing shipping rate ids to keep on incrementing.

In the new implementation, we selectively delete the old ones, and upsert the new ones. This prevents the shipping rate ids from keep on incrementing.

ecgan committed:
Delete unused ShippingRateSetup and child components.

ecgan committed:
Fix groupShippingRatesByPriceCurrency test.

ecgan committed:
Remove unneeded rates in groupShippingRatesByPriceCurrency.

ecgan committed:
Fix useShippingRatesSuggestions test based on shipping rates suggestions API changes.

The API changes are introduced in https://github.com/woocommerce/google-listings-and-ads/pull/1202.

ecgan committed:
Fix failed shipping rates reducer tests.

ecgan committed:
Fix CSS linting.

ecgan committed:
Update shipping rates schema in reducer test.

ecgan committed:
Fix deleteShippingRates in useAutoClearShippingEffect.

ecgan committed:
Check for ids in deleteShippingRates.

This is to be more defensive and prevent API error response when sending empty ids array.

ecgan committed:
Save form values when users click on Continue button.

ecgan committed:
Code refactor with isSameShippingRate.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Fix shipping rates not saved correctly in Edit Free Listings.

When users delete shipping rates, click on save changes button, and reload the page, the shipping rates would show up again and did not get deleted. This is because we were calling upsertShippingRates, and there is no delete.

ecgan committed:
Make ShippingRateSection shared component and use in Setup MC and Edit Free Listings.

ecgan committed:
Move FlatShippingRatesInputCards into ShippingRateSection component folder.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Standardize on saveShippingRates, remove upsertShippingRates and deleteShippingRates.

ecgan committed:
Use RECEIVE_SHIPPING_RATES, as SAVE_SHIPPING_RATES is not necessary.

ecgan committed:
Code refactor with SHIPPING_RATE_METHOD.FLAT_RATE.

ecgan committed:
Defensive check on shipping rate flat_rate method.

ecgan committed:
Change groupShippingRatesByPriceCurrency to also group by shipping rate method.

ecgan committed:
Rename groupShippingRatesByPriceCurrency to groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Fix error with missing shipping rate method.

The error happens when we go to EditRateModal and add a country and save the changes.

ecgan committed:
Standardize the schema in countriesPriceArray.

ecgan committed:
Change and rename groupShippingRatesByFreeShippingThreshold to also group by method.

ecgan committed:
Fix checkErrors shippingRates value should come from form value.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/edit-free-listings' into feature/shipping-rates-ui/setup-mc

Conflicts:
	js/src/components/flat-shipping-rates-input-cards/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByMethodFreeShippingThreshold.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/add-rate-button/add-rate-modal/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/countries-price-input/edit-rate-button/edit-rate-modal/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/index.scss
	js/src/setup-mc/setup-stepper/setup-free-listings/useSaveSuggestions.js

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/useAutoSaveShippingRatesEffect.js:
```diff
@@ -0,0 +1,15 @@
+import useIsEqualRefValue from '.~/hooks/useIsEqualRefValue';
+import useDebouncedCallbackEffect from '.~/hooks/useDebouncedCallbackEffect';
+
+const useAutoSaveShippingRatesEffect = ( shippingRates ) => {
+	const { saveShippingRates } = useAppDispatch();
+	const shippingRatesRefValue = useIsEqualRefValue( shippingRates );
+
+	useDebouncedCallbackEffect( shippingRatesRefValue, saveShippingRates );
+};
```
❓
After using the new `<ShippingRateSection>`, the change of shipping rates only is triggered when onBlur, what is the reason for still using debounce here? It could be simply handled by using the `onChange` prop of `<Form>`:

```js
const handleChange = ( { name, value: shippingRates } ) => {
	if ( name === 'shipping_country_rates' ) {
		saveShippingRates( shippingRates );
	}
};

// ... omitted

<Form
	// ... omitted
	onChange={ handleChange }

```

ecgan replied:
I was thinking to make it consistent with others like `useAutoSaveSettingsEffect` and `useAutoClearShippingEffect`, i.e. "auto save" will happen with a debounce delay.

eason9487 replied:
Thanks for the explanation!

I thought the only purpose of those debounces was to avoid short intervals between calls to the API, but when that concern doesn't exist, is there a need to maintain consistency for "they are all debounced"? Sorry, I'm not sure what the benefit of this is.

As the result of this implementation is workable, I think this thread is a 💅  thing now.

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/useShippingRatesSuggestions.test.js:
```diff
@@ -49,35 +49,41 @@ describe( 'useShippingRatesSuggestions', () => {
+					method: 'flat_rate',
+					currency: 'US',
+					rate: 12,
+					options: [],
```
Note: Related to the initial values mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r816551941

ecgan replied:
Updated in [`e096514` (#1267)](https://github.com/woocommerce/google-listings-and-ads/pull/1267/commits/e096514bb80c935ae342893d56cd8bb0a790e6aa).

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/index.js:
```diff
@@ -59,8 +70,40 @@ const SetupFreeListings = ( props ) => {
-	const handleSubmitCallback = () => {
-		onContinue();
+	const handleSubmitCallback = async ( values ) => {
+		/**
+		 * Even though we already have auto-save effects in the FormContent,
+		 * we are saving the form values here again to be sure,
+		 * because the auto-save may not be fired
+		 * when users were having focus on the text input fields
+		 * and then immediately click on the Continue button.
+		 */
```
💅 
This detail is very thoughtful, but I doubt merchants would easily get stuck in this situation and cannot continue. Even the free listings edit page is always like this from the beginning. Adding this code as an exchange for that processing seems a bit unworthy.

ecgan replied:
> Even the free listings edit page is always like _this_ from the beginning.

Sorry, what do you mean by "this"? I don't get it. 😅 

In Edit Free Listings, we click on the "Save changes" button, the button will show a loading icon, and we do the saving in the background.

In this Setup MC here, I think it is good to save upon users' action on clicking the button. Even though we already have the "auto-save" feature, I'd say "auto-save" is a "nice-to-have" feature, it can be removed anytime if we want, and it may fail silently when it fails. On the other hand, "save upon button click" is a "must-have" feature, and when it fails, we must inform users and we cannot proceed to the next step.

eason9487 replied:
> what do you mean by "this"? I don't get it.

Sorry, I misunderstood it as another UI detail that is not a problem actually. Please ignore that comment. 🙏 

---

> On the other hand, "save upon button click" is a "must-have" feature, and when it fails, we must inform users and we cannot proceed to the next step.

If this is a concern, then the shipping time seems to be missed in this handling?

ecgan replied:
> If this is a concern, then the shipping time seems to be missed in this handling?

Ahh yes. I think we should also do that one day, when we work on the tech debt issue and move the shipping times data into form.

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/index.js:
```diff
@@ -59,8 +70,40 @@ const SetupFreeListings = ( props ) => {
+				saveShippingRates( shippingRatesValue ),
+			] );
+			onContinue();
+		} catch ( error ) {
+			createNotice(
+				'error',
+				__(
+					'There is a problem in saving your settings and shipping rates. Please try again later.',
+					'google-listings-and-ads'
+				)
+			);
+		}
```
💅 
It's almost impossible to enter this catch block because `saveSettings` and `saveShippingRates` already catch all errors inside.

ecgan replied:
> `saveSettings` and `saveShippingRates` already catch all errors inside.

Ahh right. I'm leaning towards keeping this intact, and feeling like one day we should remove the `handleFetchError` in `actions.js`, or at least let the error propagate up to components so that the components are able to handle the error in the component if they want to. 📜 WDYT?

eason9487 replied:
Thanks for the explanation!

> WDYT?

I prefer to see how to handle error catching in the future when the error handling in *.~/data* is adjusted, otherwise when reading the code devs might wonder why there is a catch that almost never works, or spend time thinking whether there is an edge case. Or, at least add a code comment for the reason why the `catch` is here but does not work.

ecgan replied:
> I prefer to see how to handle error catching in the future when the error handling in .~/data is adjusted

I gave a few more thoughts on this and I thought of an idea: I can look into removing the try/catch block for `saveShippingRates` (refactor it with `deleteShippingRates` and `upsertShippingRates`) in actions.js in https://github.com/woocommerce/google-listings-and-ads/issues/1287, so that when there is an error in those shipping rates operations, the error would bubble up and can be handled by the components. The error will be caught by the catch block we are discussing here. Does that sound good? 🙂

eason9487 replied:
Sounds great! Thanks!

eason9487 commented on js/src/setup-mc/setup-stepper/setup-free-listings/index.js:
```diff
@@ -62,8 +70,40 @@ const SetupFreeListings = ( props ) => {
+			);
+		}
+
+		if ( isMounted() ) {
```
❔  I'm curious about what this deals with?

ecgan replied:
If the users go to other page (e.g. by clicking on the step number at the top) while the saving is in progress, there will be a console error like "setState() on an unmounted component", so we check for `isMounted()` before calling setState after all the async save operations.

eason9487 replied:
Got it. Thanks!

💅 
Since the API error won't catch in this `handleSubmitCallback`, no matter whether an error occurs or not in *actions.js*, `onContinue()` will be triggered and proceed to the next step, so perhaps there is no need to `setSaving( false );` here.

ecgan replied:
> Since the API error won't catch in this `handleSubmitCallback`, no matter whether an error occurs or not in _actions.js_, `onContinue()` will be triggered and proceed to the next step

Well, clearly this is an issue. 😅 

Related to https://github.com/woocommerce/google-listings-and-ads/pull/1267#discussion_r820908402, to address the above issue, I'm thinking to remove the try/catch block in `saveShippingRate` in actions.js in https://github.com/woocommerce/google-listings-and-ads/issues/1287, so that the error would bubble up and can be caught by the component. With that, the `setSaving( false )` would be possible and meaningful. Does that sound good?

eason9487 replied:
Sounds good!

A further thought is, if that API error would be thrown to this component, then perhaps we could move `setSaving( false )` into `catch` block. The continue button could be set back to enabled only if errors occurred, otherwise, keeping it showing the loading state before bringing merchant to the next step should be fine.

eason9487 reviewed (commented):
❔ 
When switching the option of "I offer free shipping for orders over a certain price" from yes to no, it will clear all filled-in free shipping thresholds. But If change the shipping rate of countries, which have a filled-in free shipping threshold, to zero and then change to non-zero again, the filled-in free shipping threshold is still there. Is this inconsistency to be expected?

https://user-images.githubusercontent.com/17420811/156556576-0dc4f798-c036-46d9-af9e-4f701e25af9d.mp4

ecgan committed:
Merge branch 'feature/shipping-rates-ui/edit-free-listings' into feature/shipping-rates-ui/setup-mc

Conflicts:
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByMethodFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate-section.js

ecgan committed:
SImplify code in groupShippingRatesByMethodFreeShippingThreshold.

ecgan committed:
Change options from [] to {}.

This is as per the changes in https://github.com/woocommerce/google-listings-and-ads/pull/1284.

ecgan commented:
@eason9487,

> If change the shipping rate of countries, which have a filled-in free shipping threshold, to zero and then change to non-zero again, the filled-in free shipping threshold is still there. Is this inconsistency to be expected?

Thanks, good catch! The free shipping threshold should be removed. I've created issue https://github.com/woocommerce/google-listings-and-ads/issues/1291 for this. 👍

ecgan committed:
Fix code comment in MinimumOrderCard.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js:
```diff
@@ -98,14 +98,14 @@ const MinimumOrderCard = ( props ) => {
-		 * we render the groups with defined threshold first.
-		 * Then, if there is a group with undefined threshold,
-		 * we render an "Add another minimum order" button last.
+		 * The group with undefined threshold. This is used
+		 * to render the "Add another minimum order" button
+		 * after all the groups with defined threshold.
```
Thanks for the update! However, it still looks a bit imprecise. Would it be much closer to this description?
- The group with undefined threshold could be more than one when there are multiple currencies in the shipping rates.
- The first group (`emptyThresholdGroup`) with undefined threshold would be rendered at the bottom with "Add another minimum order" button.
- The `thresholdGroups` are the groups with defined or undefined threshold, except for the `emptyThresholdGroup` one.

ecgan replied:
> * The group with undefined threshold **_could be more than one_** when there are **_multiple currencies_** in the shipping rates.
> * The `thresholdGroups` are **_the groups with defined or undefined threshold_**, except for the `emptyThresholdGroup` one.

Hey @eason9487 , thanks, I just realized this. I wanted the `thresholdGroups` to be groups with defined threshold only, to render `MinimumOrderInputControl`. I'll fix this.

ecgan replied:
Fixed in [`e6623ed` (#1267)](https://github.com/woocommerce/google-listings-and-ads/pull/1267/commits/e6623eddf2b46f40338896e35138d91abc34f8ec).

eason9487 replied:
🚧  or 📜  I found multi-currency becomes a complex problem here: 😂 

1. Only the first undefined threshold can be added. When the country to be set is the latter, the UI becomes incomprehensible for merchants.

   https://user-images.githubusercontent.com/17420811/157227874-6a0d9fc0-e81f-47be-abc5-a9198ad07689.mp4

2. It's possible to edit countries' thresholds across multi-currency. Although the currencies won't be overwritten, the results look a bit surprising.

   https://user-images.githubusercontent.com/17420811/157228468-895c73bb-8f83-40ff-9d96-fa68123fe959.mp4

3. It's possible to edit countries' shipping rates across multi-currency. The currencies could be overwritten, but the results look much more surprising when comparing this to 2.

   https://user-images.githubusercontent.com/17420811/157228702-2714f975-1599-4aea-a3e8-76d00f55aa4d.mp4

But this is a rare edge case and probably needs more discussion (and maybe UI design change), so I think probably we could create a follow-up issue to see how it could be solved later.

ecgan replied:
LOL, it does. 😂 

For the issue 1, I think it is still okay. It may be tedious and difficult, but generally the final outcome is still correct.

For issue 2, that does look buggy, because users are putting "234 **AUD**" for US in the minimum order modal, but end up showing "234 **USD**" for US in the final outcome in the page. I'm thinking that can be solved if we have separate currency for the free shipping threshold, e.g.:

```js
const shippingRate = {
	id: '1',
	country: 'US',
	method: 'flat_rate',
	currency: 'USD', // currency for shipping rate.
	rate: 4.99,
	options: {
		free_shipping_threshold: 88,
		free_shipping_threshold_currency: 'AUD', // currency for free shipping threshold.
	},
}
```

However, the above idea may be too much though. 

Another idea is to simply restrict the country options. In the modal, we will only show country options that has the same AUD shipping rate currency, so country US won't show up in the drop down. The "downside" is users may be wondering for a bit "why is there no US in the drop down?" (answer: because the currency doesn't match). 

I'll create a separate issue for this. 📜 

For issue 3, although it looks a bit weird, I think that is still okay, because what the users put in the modal matches the final outcome in the page.

ecgan replied:
> I'll create a separate issue for this. 📜

Note: I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1324 for the issue 2.

eason9487 reviewed (approved):
Tested locally and works. Left a comment about the code comment update https://github.com/woocommerce/google-listings-and-ads/pull/1267/commits/d6b479808cfc5e8727699f23e77c054c19b93ee5.

There are some readability concerns but don't block this PR. They could be ignored or be addressed in subsequence PR. I think this PR is ready to move on.

💡 On a side note, a part of [shipping values flash issue](https://github.com/woocommerce/google-listings-and-ads/issues/1078) is also addressed with this PR. I think this can be referred to when handling the shipping times part someday. 👍

ecgan committed:
Fix thresholdGroups containing groups with undefined threshold.

ecgan commented:
> 💡 On a side note, a part of [shipping values flash issue](https://github.com/woocommerce/google-listings-and-ads/issues/1078) is also addressed with this PR. I think this can be referred to when handling the shipping times part someday. 👍

Oh yes, that's the idea with all the huge changes in this shipping rates UI change project. I'm mentioning the issue number https://github.com/woocommerce/google-listings-and-ads/issues/1078 directly here so that this PR would show up in the issue page as well. 😄
